### PR TITLE
Issue 811: PyPi controller

### LINKED ISSUE #811: Implement a PypiArtifactController
# Task Description

We need to:
* [ ] Implement a controller that handles incoming requests for PyPi packages
* [ ] Investigate what PyPi commands, (if any), we need to support

The code should be located in a class called `PypiArtifactController` under `strongbox/strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/layout/pypi`.

# Tasks

* [ ] Implement downloading of artifacts
  * [ ] Implement support for the `data-requires-python` attribute as described in [PEP 503](https://www.python.org/dev/peps/pep-0503/)
* [ ] Implement uploading of artifacts
* [ ] Fix the browsing and downloading of PyPi artifacts

# Task Relationships

This is a sub-task of #807. It should be relatively straight-forward to implement this and you should get acquainted with the master task, before proceeding.

This task depends on #1190.

# See Also

* pypa/pipenv#2075 : Allow users to override default PyPI index URL with PyPI mirror URLs (without modifying Pipfile)
* [Mulit-part handling](https://stackoverflow.com/questions/55894007/how-to-read-multipart-using-sring-boot/)
* [Browseable PyPi directory structure](https://mirrors.aliyun.com/pypi/simple/)

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @fuss86

### PR DESCRIPTION
**Pull request by grantfar**
# Pull Request Description

This pull request closes #811 .

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [ ] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [ ] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [ ] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [x] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [ ] No

### CODE REVIEW

**Comment by carlspring** [comment]
Thanks for this contribution!

Would you mind creating a task list in the pull request's description where you outline what tasks you've done so far and what needs to still be done...? Thanks! :)

Also, could you please make sure you've formatted the code according to our convention? Thanks! :)

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/layout/pypi/PypiArtifactController.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+                                      String Filename)
+    {
+        StringBuilder builder = new StringBuilder(Name);
+
+        builder.append("-").append(Version).append(FileType.equals("bdist_wheel") ? ".whl":".tar.gz");
+
+        return Filename.equals(builder.toString());
+    }
+
+    private ResponseEntity<String> upload(Repository repository,
```
Comment (kaansonmezoz) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/layout/pypi/PypiArtifactController.java]: This logic should not be implemented in controller. Controllers should be only contains interaction between request service layers and response. So uploading logic should be extracted to somewhere else. This way controller will be cleaner, more neat

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/layout/pypi/PypiArtifactController.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        {
+            logger.error(e.getMessage(), e);
+
+            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(e.getMessage());
+        }
+
+    }
+
+
+    private ResponseEntity<String> delete(Repository repository,
```
Comment (kaansonmezoz) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/controllers/layout/pypi/PypiArtifactController.java]: Should be extracted to service layer

**Comment by carlspring** [comment]
Thanks for reviewing @kaansonmezoz , but this pull request should have been closed a long time ago, as a competing one already covered this functionality.

**Pull request closed without merging**